src/main/java/htsjdk/samtools/util/IOUtil.java

```diff
@@ -988,4 +988,45 @@ public static Path getPath(String uriString) throws IOException {
             return FileSystems.newFileSystem(uri, new HashMap<>(), cl).provider().getPath(uri);
         }
     }
+    
+    /**
+     * Test whether a input stream looks like a GZIP input.
+     * @param bufferedinput the input stream. The buffer must be large enough to contain the GZIP signature
+     * @return true if `bufferedinput` starts with a gzip signature 
+     * @throws IOException
+     */
+    public static boolean isGZIPInputStream(final BufferedInputStream bufferedinput) throws IOException {
+        if(bufferedinput == null) throw new IllegalArgumentException("isGZIPInputStream(null)");
+        // see http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped
+        final byte[] signature = new byte[2];
+        bufferedinput.mark(signature.length);
+        final int len = bufferedinput.read(signature); // read the signature
+        bufferedinput.reset(); // push back the signature to the stream
+        return  len == signature.length && 
+                    signature[0] == (byte) 0x1f && 
+                    signature[1] == (byte) 0x8b; 
+    }
+    
+    /**
+     * If `in` is a GZipped stream, wrap it into a {@link GZIPInputStream}.
+     * @param in the input stream. If it's a BufferedInputStream, the buffer must be large enough to contains the gzip signature
+     * @return a decompressed input stream
+     * @throws IOException
+     */
+    public static InputStream mayBeGZippedInputStream(final InputStream in) throws IOException {
+        if(in == null) throw new IllegalArgumentException("mayBeGZippedInputStream(null)");
+        /* BufferedInputStream is needed to decode the first bytes of the stream */
+        final BufferedInputStream bufferedinput;
+        /* it's already a buffered input stream, no need to wrap it again */
+        if(in instanceof BufferedInputStream) {
+            bufferedinput = BufferedInputStream.class.cast(in);
```

Comment:
@magicDGS I'm currently looking at this. Unless I'm misunderstanding there is no way to return the original InputStream because you need to consume the first bytes of the stream.